blog: Introducing secrets (#540)
wasmCloud/wasmCloud.com: blog/2024-08-07-introducing-secrets-how-wasmcloud-solves-for-secrets-in-distributed-environments/index.mdx

* The request goes to a **secrets handler** which in turn requests the specified secret from the appropriate **secrets backend**.
* Communication with a **secrets store** is mediated by a **secrets backend**.

You can see the path of a secret request for a component in the diagram below. In this example, we can see how the system integrates with both an external secret store like Vault and the key-value store already available in the wasmCloud stack as part of NATS.

![diagram of secret flow when starting component with secret](../../static/img/start-component-with-secret.png)

Let's walk through this process step-by-step:

* The `wash` CLI asks the wasmCloud host to start a component with `SECRET_foo`
* The host fetches a secret reference from the NATS key-value bucket
* The NATS key-value bucket returns the secret reference to the host
* The host fetches a secret named `foo` from the secrets backend
* The secrets backend authorizes a secret fetch by the component JWT
* The secrets backend returns the encrypted value of secret `foo` to the host

> * The host starts the component, storing the secret inthe host handler using the `secrecy` crate

Reviewer: typo: `in the`

## How secrets work in wasmCloud

Secrets are a challenging problem in cloud native environments, so we spent a lot of time thinking about how to approach them in wasmCloud.  

We wanted a better approach&mdash;one that leverages the NATS-powered lattice to deliver true encryption and avoid storing secrets on the [wasmCloud host](/docs/concepts/hosts) in any form that would be accessible to anything other than the component or provider requesting them. 

We also wanted our secret system to be **pluggable**, so teams could use the secret store of their choice, whether that was HashiCorp Vault, cloud provider systems like AWS Secrets Manager, an in-house solution, or an open source store.

To achieve all of this, we came up with an architecture that organizes secret requests and responses into a few key steps:

> * A component or capability provider **requests a secret** by reference to secret's location, without needing to present credential material. 

Reviewer 2: ```suggestion
* A component or capability provider **requests a secret** by reference to secret's location in the secret store, without needing to present credential material. 
```

* The request goes to a **secrets handler** which in turn requests the specified secret from the appropriate **secrets backend**.
* Communication with a **secrets store** is mediated by a **secrets backend**.

You can see how a component requests a secret in the diagram below:

![diagram of secret flow when starting component with secret](../../static/img/secrets-diagram.png)

If we zoom into an individual component level, the standard flow for a secret looks like this:

![diagram](./images/component-secret.png)

Let's walk through this process step-by-step:

1. Component sends `wasmcloud:secrets/get` call to host
2. Host returns secret reference
3. Component sends `wasmcloud:secrets/reveal` call to host
4. Host exposes secret from `secrecy` crate's in-memory store, returning value to component

> The secret itself is never logged or accessible in memory&mdash;it is encrypted with an [**xkey**](https://docs.nats.io/running-a-nats-service/configuration/securing_nats/auth_callout#encryption), a x25519 keypair, and held in-memory with Rust's [`secrecy` crate](https://crates.io/crates/secrecy). When you drop a secret, it is wiped in-memory. 

Reviewer 2: ```suggestion
The secret itself is never logged or accessible in memory&mdash; while in transit, the secret is encrypted with an [**xkey**](https://docs.nats.io/running-a-nats-service/configuration/securing_nats/auth_callout#encryption), a x25519 keypair, and when held in-memory by the Host, the host uses the Rust [`secrecy` crate](https://crates.io/crates/secrecy) to secure the contents. The secret is never stored on disk by the host and once you stop needing the secret, it is wiped from the Host's memory. 
```

Reviewer 2: This might be nuance that's not relevant for the blog post, but I figured I'd offer it in case it felt like it adds to the description.

Using a common API enables the same vendor-agnostic approach that we take throughout the rest of the wasmCloud ecosystem&mdash;if you need to swap out your secrets vendor, that's not a problem, as your components and providers remain unchanged while leveraging exactly the same API. All you need to do is change one or two lines of configuration. 

At the launch of secrets support, backends are available for:

* [HashiCorp Vault](https://github.com/wasmCloud/wasmCloud-contrib/tree/main/secrets/secrets-vault)
* [NATS KV store](https://github.com/wasmCloud/wasmCloud/tree/main/crates/secrets-nats-kv)
* [Kubernetes Secrets](https://github.com/wasmCloud/wasmCloud-contrib/tree/main/secrets/secrets-kubernetes)

> Providing support for Kubernetes Secrets is particularly powerful, since it means that teams [running wasmCloud on Kubernetes](/blog/how-to-run-webassembly-components-on-kubernetes-with-wasmcloud) can immediately use their existing secrets systems with wasmCloud applications. 

Reviewer 2: I figured it might be worth illustrating why this is so powerful beyond just the Kubernetes Secrets themselves:
```suggestion
Providing support for Kubernetes Secrets is particularly powerful, since it means that teams [running wasmCloud on Kubernetes](/blog/how-to-run-webassembly-components-on-kubernetes-with-wasmcloud) can immediately use their existing secrets systems with wasmCloud applications. This would even allow you to do things like extending your wasmCloud application secrets to be pulled in via Kubernetes Secrets that are provisioned by an external process such as [External Secrets Operator](https://external-secrets.io/latest/).
```
